Include/cpython/pyatomic_msc.h

```diff
@@ -0,0 +1,860 @@
+#ifndef Py_ATOMIC_MSC_H
+#  error "this header file must not be included directly"
+#endif
+
+// This is the implementation of Python atomic operations for MSVC if the
+// compiler does not support C11 or C++11 atomics.
+
+#include <intrin.h>
+
+
+static inline int
+_Py_atomic_add_int(int *address, int value)
+{
+    return (int)_InterlockedExchangeAdd((volatile long*)address, (long)value);
+}
+
+static inline int8_t
+_Py_atomic_add_int8(int8_t *address, int8_t value)
+{
+    return (int8_t)_InterlockedExchangeAdd8((volatile char*)address, (char)value);
+}
+
+static inline int16_t
+_Py_atomic_add_int16(int16_t *address, int16_t value)
+{
+    return (int16_t)_InterlockedExchangeAdd16((volatile short*)address, (short)value);
+}
+
+static inline int32_t
+_Py_atomic_add_int32(int32_t *address, int32_t value)
+{
+    return (int32_t)_InterlockedExchangeAdd((volatile long*)address, (long)value);
+}
+
+static inline int64_t
+_Py_atomic_add_int64(int64_t *address, int64_t value)
+{
+#if defined(_M_X64) || defined(_M_ARM64)
+    return (int64_t)_InterlockedExchangeAdd64((volatile __int64*)address, (__int64)value);
+#else
+    for (;;) {
+        __int64 old_value = *(volatile __int64*)address;
+        __int64 new_value = old_value + (__int64)value;
+        if (old_value == _InterlockedCompareExchange64((volatile __int64*)address, new_value, old_value)) {
+            return old_value;
+        }
+    }
+#endif
+}
+
+static inline intptr_t
+_Py_atomic_add_intptr(intptr_t *address, intptr_t value)
+{
+#if SIZEOF_VOID_P == 8
+    return (intptr_t)_InterlockedExchangeAdd64((volatile __int64*)address, (__int64)value);
+#else
+    return (intptr_t)_InterlockedExchangeAdd((volatile long*)address, (long)value);
+#endif
+}
+
+static inline unsigned int
+_Py_atomic_add_uint(unsigned int *address, unsigned int value)
+{
+    return (unsigned int)_InterlockedExchangeAdd((volatile long*)address, (long)value);
+}
+
+static inline uint8_t
+_Py_atomic_add_uint8(uint8_t *address, uint8_t value)
+{
+    return (uint8_t)_InterlockedExchangeAdd8((volatile char*)address, (char)value);
+}
+
+static inline uint16_t
+_Py_atomic_add_uint16(uint16_t *address, uint16_t value)
+{
+    return (uint16_t)_InterlockedExchangeAdd16((volatile short*)address, (short)value);
+}
+
+static inline uint32_t
+_Py_atomic_add_uint32(uint32_t *address, uint32_t value)
+{
+    return (uint32_t)_InterlockedExchangeAdd((volatile long*)address, (long)value);
+}
+
+static inline uint64_t
+_Py_atomic_add_uint64(uint64_t *address, uint64_t value)
+{
+    return (uint64_t)_Py_atomic_add_int64((volatile int64_t*)address, (int64_t)value);
```

Comment:
There is a compiler warning: ``different 'volatile' qualifiers``. _Py_atomic_add_int64() doesn't need volatile.